### PR TITLE
Pw2wann workchain

### PR DESCRIPTION
**Pull request by DanielMarchand**
Add a SplitPw2wann WorkChain to help parallelize that step

### CODE REVIEW

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+                mmn_collected[:, :, bnd, newindex_bandgroup] = mmn_i[:, :,
+                                                                     j, :]
+
+        # prep the output folder
+        folder = FolderData()
+
+        # Writing mmn file, unfortunately this is a bit messy because
+        # write_mmn dumps to a file which must be read again later
+        mmn_filename = "aiida.mmn"
+        with tempfile.TemporaryDirectory() as tmp_dir:
```
Comment (greschd) [inline aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: AFAICT we only use a single file from the `TemporaryDirectory` here - meaning we could also use `NamedTemporaryFile` instead.

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+            output_file = os.path.join(tmp_dir, mmn_filename)
+            write_mmn(output_file, mmn_collected, kpb_kidx, kpb_g)
+
+            # mmn_collected might be very large so we want to remove
+            # it from memory before reading in a copy
+            del mmn_collected
+            gc.collect()  # for paranoia's sake run garbage collection
+
+            with open(output_file, mode='r') as f_in:
```
Comment (greschd) [inline aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: Here we could use `FolderData.put_object_from_file` to copy the file (probably, that'll be more efficient).

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+                with folder.open(mmn_filename, mode='w') as f_out:
+                    f_out.write(f_in.read())
+
+        # writing amn, eig (and any other) outputs
```
Comment (greschd) [inline aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: I think the `(and any other)` part of this comment should be a TODO: Currently, other files are _not_ copied. 

We'll probably also need some logic to determine _which_ files should be retrieved.

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+        filenames = ['aiida.eig', 'aiida.amn']
+        no_mmn_calc = self.ctx['pw2wann_no_mmn']
+        r_folder = no_mmn_calc.outputs.remote_folder
+        for filename in filenames:
+            with tempfile.TemporaryDirectory() as tmp_dir:
```
Comment (greschd) [inline aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: Here, I'm not sure if we should put all the files into the same `TemporaryDirectory` (switching the loop order), and then use `FolderData.put_object_from_tree` to copy the whole directory to the repository at once.

That'd use more disk space (temporarily), but IMO the code would be a bit simpler.

**Review by greschd — COMMENTED** [review]
Thanks @DanielMarchand, looks good overall - I added some minor comments / suggestions.

I haven't yet ran the tests locally, but will do so to check if there are any changes needed there (on the QE part that doesn't run in CI). On a related note, I'm not sure why CI didn't run at all here...

**Comment by greschd** [comment]
Regarding the merge conflicts: I'm fairly certain that `exclude=['pw.structure', 'kpoints']` (the version in `develop`) is the correct one.

For the second merge conflict (from the `UNK*` PR https://github.com/greschd/aiida-tbextraction/pull/19) you can just add that to the TODO comment for retrieving additional files from the "no MMN" calculation.

**Comment by greschd** [comment]
> On a related note, I'm not sure why CI didn't run at all here...

I took the liberty to rebase the commits, to make the CI run. The problem was that this branch didn't contain the changes to move from Travis to Github Actions.

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+# © 2017-2019, ETH Zurich, Institut für Theoretische Physik
+# Author: Dominik Gresch <greschd@gmx.ch>
+"""
+Defines a workflow that calculates the Wannier90 input files using Quantum ESPRESSO pw.x.
+"""
+
+import copy
+import gc
+import io
+import more_itertools
```
Comment (greschd) [inline aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: `more_itertools` is not in the standard library, this belongs in the second block of imports.

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+import re
+import tempfile
+import os
+
+import numpy as np
+from w90utils.io import write_mmn
+
+from collections import ChainMap
```
Comment (greschd) [inline aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: `collections` _is_ in the standard library, this belongs in the first block of imports.

**Comment by greschd** [comment]
I think the tests also need some changes, for me `tests/fp_run/test_qe_fp_run.py` fails.

File: tests/conftest.py [hunk 7]
```diff
@@ -263,8 +263,12 @@ def inner():
                     'metadata': get_metadata_singlecore()
                 },
                 'pw2wannier': {
-                    'code': code_pw2wannier90,
-                    'metadata': get_metadata_singlecore()
+                    # Use a huge batchsize to force it all at once
```
Comment (greschd) [inline tests/conftest.py]: Shouldn't there be at least one test where we actually use more than one batch?

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py (hunk 3 above)
Comment (DanielMarchand) [reply aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: Correct, let's discuss

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py (hunk 1 above)
Comment (DanielMarchand) [reply aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: added in latest commit

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py (hunk 2 above)
Comment (DanielMarchand) [reply aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: added in latest commit

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py (hunk 5 above)
Comment (DanielMarchand) [reply aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: corrected in latest commit

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py (hunk 6 above)
Comment (DanielMarchand) [reply aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: corrected in latest commit

File: tests/conftest.py (hunk 7 above)
Comment (DanielMarchand) [reply tests/conftest.py]: Yeah that would make sense

**Comment by DanielMarchand** [comment]
> I think the tests also need some changes, for me `tests/fp_run/test_qe_fp_run.py` fails.
Working again

File: aiida_tbextraction/fp_run/wannier_input/_qe.py [hunk 8]
```diff
@@ -13,7 +13,7 @@
 from aiida_tools import check_workchain_step
 from aiida_wannier90.calculations import Wannier90Calculation
 from aiida_quantumespresso.workflows.pw.base import PwBaseWorkChain
-from aiida_quantumespresso.calculations.pw2wannier90 import Pw2wannier90Calculation
+from aiida_tbextraction.fp_run.wannier_input.pw2wann import SplitPw2wannier90
```
Comment (greschd) [inline aiida_tbextraction/fp_run/wannier_input/_qe.py]: For consistency, we should name the file `pw2wannier` (sorry, didn't spot that one last time around).

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py (hunk 3 above)
Comment (greschd) [reply aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: I don't think this needs doing right now, just adding a TODO comment stating what is currently retrieved and that other output files aren't is enough.

File: aiida_tbextraction/fp_run/wannier_input/pw2wann.py (hunk 4 above)
Comment (greschd) [reply aiida_tbextraction/fp_run/wannier_input/pw2wann.py]: @DanielMarchand opinions on this? We can also leave it, up to you.

File: tests/conftest.py (hunk 7 above)
Comment (greschd) [reply tests/conftest.py]: Fixed in 598e5d1

**Pull request merged**